fe/fe-core/src/test/java/com/starrocks/sql/plan/SubqueryTest.java

```diff
@@ -569,4 +569,285 @@ public void testCorrelatedScalarNonAggSubqueryWithExpression() throws Exception
+            String plan = getFragmentPlan(sql);
+            assertContains(plan, "  8:NESTLOOP JOIN\n" +
+                    "  |  join op: CROSS JOIN\n" +
+                    "  |  colocate: false, reason: \n" +
+                    "  |  other predicates: 11: expr + 3: v3 = 5: v5");
+            assertContains(plan, "  4:HASH JOIN\n" +
+                    "  |  join op: LEFT OUTER JOIN (BROADCAST)\n" +
+                    "  |  colocate: false, reason: \n" +
+                    "  |  equal join conjunct: 5: v5 = 8: v2");
+        }
+        {
+            String sql = "select * from t0 " +
+                    "join t1 on t0.v1 = (select count(*) from t1 where t0.v2 = t1.v5 and t0.v3 = t1.v6)";
+            String plan = getFragmentPlan(sql);
+            assertContains(plan, "  8:NESTLOOP JOIN\n" +
+                    "  |  join op: CROSS JOIN\n" +
+                    "  |  colocate: false, reason: ");
+            assertContains(plan, "  4:HASH JOIN\n" +
+                    "  |  join op: LEFT OUTER JOIN (BROADCAST)\n" +
+                    "  |  colocate: false, reason: \n" +
+                    "  |  equal join conjunct: 2: v2 = 8: v5\n" +
+                    "  |  equal join conjunct: 3: v3 = 9: v6\n" +
+                    "  |  other predicates: 1: v1 = ifnull(10: count, 0)");
+        }
+    }
+
+    @Test
+    public void testOnClauseCorrelatedScalarNonAggSubquery() throws Exception {
+        {
+            String sql = "select * from t0 " +
+                    "join t1 on t0.v1 = (select v4 from t1 where t0.v2 = t1.v5)";
+            String plan = getFragmentPlan(sql);
+            assertContains(plan, "  10:NESTLOOP JOIN\n" +
+                    "  |  join op: CROSS JOIN\n" +
+                    "  |  colocate: false, reason: ");
+            assertContains(plan, "  4:HASH JOIN\n" +
+                    "  |  join op: LEFT OUTER JOIN (BROADCAST)\n" +
+                    "  |  colocate: false, reason: \n" +
+                    "  |  equal join conjunct: 2: v2 = 8: v5");
+        }
+        {
+            String sql = "select * from t0 " +
+                    "join t1 on t0.v1 = t1.v4 " +
+                    "and t0.v3 = (select v4 from t1 where t0.v2 = t1.v5)";
+            String plan = getFragmentPlan(sql);
+            assertContains(plan, "  10:HASH JOIN\n" +
+                    "  |  join op: INNER JOIN (BROADCAST)\n" +
+                    "  |  colocate: false, reason: \n" +
+                    "  |  equal join conjunct: 1: v1 = 4: v4");
+            assertContains(plan, "  4:HASH JOIN\n" +
+                    "  |  join op: LEFT OUTER JOIN (BROADCAST)\n" +
+                    "  |  colocate: false, reason: \n" +
+                    "  |  equal join conjunct: 2: v2 = 8: v5");
+        }
+        {
+            String sql = "select * from t0 " +
+                    "join t1 on t0.v1 = (select v4 from t1 where t0.v2 = t1.v5 and t1.v5 = 10 and t1.v6 != 3)";
```

Comment:
can be support complex expression? like: `join t1 on t0.v1 = (t1.v2 in (xxxxxx))` or `join t1 on t0.v1 = (exists (xxxxxx))`?

Reply:
Yes, we supporte this kind of subquery